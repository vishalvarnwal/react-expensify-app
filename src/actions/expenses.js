import uuid from 'uuid'; 

//Add Expenses
export const addExpense = (
    { 
        description = '',
        note = '', 
        amount = 0, 
        createdAt = 0 
    } = {}
    ) =>({
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            description,
            note,
            amount,
            createdAt
        }
});

//Remove Expenses
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id  
})

//Edit Expenses
export const editExpense = (id, updates) =>({
    type: 'EDIT_EXPENSE',
    id,
    updates
});