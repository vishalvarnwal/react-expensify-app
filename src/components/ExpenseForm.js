import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//const date = new date();
//const now = moment();
//console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props){
        super();
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? parseFloat(props.expense.amount/100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt
                ) : moment(),
            calendarFocused: false,
            error: undefined
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;       //if directly used e.target.value in this.setState function, it
        this.setState(() =>({ description }));    //will not work for that we need to use e.persist();
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({ amount }))
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState (() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            const error = 'Please provide description and amount.'
            this.setState(() => ({ error }))
        }else {
            this.setState(() => ({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: (this.state.createdAt).valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    className="text-inputs"
                    type = 'text'
                    placeholder = 'description'
                    autoFocus
                    value = {this.state.description}
                    onChange = {this.onDescriptionChange}
                />
                <input
                    className="text-inputs"
                    type = 'number'
                    placeholder = 'amount'
                    autoFocus
                    value = {this.state.amount}
                    onChange = {this.onAmountChange}
                />
                <SingleDatePicker
                    date = {this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.calendarFocused}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                />
                <textarea 
                    className="textarea"
                    placeholder = 'Add a note for your expense (optional)'
                    value = {this.state.note}
                    onChange = {this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className="button">Save Expense</button> 
                </div>
                   
            </form>
            
        )
    }
}