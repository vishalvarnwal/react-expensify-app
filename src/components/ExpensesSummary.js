import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import expenseTotal from '../selectors/expenses-total';

export const ExpesnesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedTotalAmount = numeral(expensesTotal/100).format('$0,0.00')

    return(
        <div>
            <h1>
                viewing {expenseCount} {expenseWord} totalling {formattedTotalAmount}
            </h1>
        </div>
    )
}

const mapStoreToProps = (state) => {
    const visibeExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibeExpenses.length,
        expensesTotal: expenseTotal(visibeExpenses)
    }
};
export default connect(mapStoreToProps)(ExpesnesSummary);