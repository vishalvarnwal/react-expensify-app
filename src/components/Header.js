import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () =>(
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>HomePage</NavLink>
        <NavLink to="/create" activeClassName="is-active">CreatePage</NavLink>
        <NavLink to="/help" activeClassName="is-active">HelpPage</NavLink>
    </header>
);

export default Header;