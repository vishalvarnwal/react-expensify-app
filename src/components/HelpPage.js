import React from 'react';

const HelpPage = () =>(
    <div>
        this is my help expense component
    </div>
)

export default HelpPage;