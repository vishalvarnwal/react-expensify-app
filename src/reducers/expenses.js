//Expense Reducer
const expenseReducerDefaultState = [];
export default (state = expenseReducerDefaultState, action)=>{
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state, action.expense
            ]

        case 'REMOVE_EXPENSE':
            //return state.filter((item)=>{return item.id !== action.id })   //without destructuring
            return state.filter(         //destructuring
                ({id})=>(
                id !== action.id
            )) 

        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if(expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense;
                }
            })

        default:
            return state;
    }
}
