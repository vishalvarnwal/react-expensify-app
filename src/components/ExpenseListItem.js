import React from 'react';
import { removeExpense} from '../actions/expenses';
import { Link } from 'react-router-dom';
 

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{amount} - {createdAt}</p>
    </div>
);



export default ExpenseListItem;