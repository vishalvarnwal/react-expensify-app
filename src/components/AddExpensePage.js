import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render(){
        return(
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit = {this.onSubmit}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return{
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    }
}
export default connect(undefined, mapDispatchToProps)(AddExpensePage);



//before testing , using this way below

// export const AddExpensePage = (props) =>(
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm 
//             onSubmit = {(expense) => {
//                 props.dispatch(addExpense(expense));
//                 props.history.push('/');               //history.push() is the property of routing
//             }}
//         />
//     </div>
// )