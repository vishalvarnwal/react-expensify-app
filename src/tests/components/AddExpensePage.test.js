import React from 'react';
import {AddExpensePage} from '../../components/AddExpensePage';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

let addExpense, history, wrapper;
// this function belongs to jest, beforeEach is called when we want to particular line of codes run for all
//  the test cases at the beginning
beforeEach(() => {
    addExpense =jest.fn();
    history = {push: jest.fn()}
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history}/>);
});

test('should render the AddExpesne Page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[1]);
})