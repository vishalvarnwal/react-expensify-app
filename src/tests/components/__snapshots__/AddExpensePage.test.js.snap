// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the AddExpesne Page correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
